src/lib/server/use_boost_future.hpp

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <boost/asio/detail/config.hpp>
+#include <memory>
+
+#include <boost/asio/detail/push_options.hpp>
+
+namespace boost {
+namespace asio {
+
+/// Class used to specify that an asynchronous operation should return a future.
+/**
+ * The use_future_t class is used to indicate that an asynchronous operation
+ * should return a std::future object. A use_future_t object may be passed as a
+ * handler to an asynchronous operation, typically using the special value @c
+ * boost::asio::use_future. For example:
+ *
+ * @code std::future<std::size_t> my_future
+ *   = my_socket.async_read_some(my_buffer, boost::asio::use_future); @endcode
+ *
+ * The initiating function (async_read_some in the above example) returns a
+ * future that will receive the result of the operation. If the operation
+ * completes with an error_code indicating failure, it is converted into a
+ * system_error and passed back to the caller via the future.
+ */
+template <typename Allocator = std::allocator<void> >
+class use_boost_future_t
+{
+ public:
+  /// The allocator type. The allocator is used when constructing the
+  /// @c std::promise object for a given asynchronous operation.
+  typedef Allocator allocator_type;
+
+  /// Construct using default-constructed allocator.
+  BOOST_ASIO_CONSTEXPR use_boost_future_t()
+  {
+  }
+
+  /// Construct using specified allocator.
+  explicit use_boost_future_t(const Allocator& allocator)
+    : allocator_(allocator)
+  {
+  }
+
+  /// Specify an alternate allocator.
+  template <typename OtherAllocator>
+  use_boost_future_t<OtherAllocator> operator[](const OtherAllocator& allocator) const
+  {
+    return use_boost_future_t<OtherAllocator>(allocator);
+  }
+
+  /// Obtain allocator.
+  allocator_type get_allocator() const
+  {
+    return allocator_;
+  }
+
+ private:
+  Allocator allocator_;
+};
+
+/// A special value, similar to std::nothrow.
+/**
+ * See the documentation for boost::asio::use_future_t for a usage example.
+ */
+#if defined(BOOST_ASIO_MSVC)
```

Comment:
we can remove this if case because we don't support windows :)

Reply:
I was wondering how closely we should align the use_boost_future.hpp to boost's use_future.hpp. Changing less would make applying later 'upstream' changes easier -- for now I've just replaced some `std::future`/`std::promise` with `boost::future`/`boost::promise`.

Reply:
okay, then just leave it in